### PR TITLE
fix permission check for creating comment while mail

### PR DESCRIPTION
**Pull request by a1012112796**
only creating comment on locked issue request write permission,
for others, read permission is enough.

related to https://github.com/go-gitea/gitea/pull/22056

/cc @KN4CK3R

### CODE REVIEW

File: services/mailer/incoming/incoming_handler.go [hunk 1]
```diff
@@ -71,11 +71,16 @@ func (h *ReplyHandler) Handle(ctx context.Context, content *MailContent, doer *u
 		return err
 	}
 
-	if !perm.CanWriteIssuesOrPulls(issue.IsPull) || issue.IsLocked && !doer.IsAdmin {
+	if !perm.CanWriteIssuesOrPulls(issue.IsPull) && issue.IsLocked && !doer.IsAdmin {
```
Comment (wolfogre) [inline services/mailer/incoming/incoming_handler.go]: I think the old code tried to do something like:

- Return nil if the doer cannot write
- Even if the doer can write, return nil if the issue is locked and the doer is not admin.

So maybe we could just change it to:

```go
if !perm.CanReadIssuesOrPulls(issue.IsPull) || issue.IsLocked && !doer.IsAdmin {
```

---

Edit:

So the old code is wrong.

Comment (lunny) [reply services/mailer/incoming/incoming_handler.go]: But with write permission, it could post comment even if issue.IsLocked from UI experience?

Comment (a1012112796) [reply services/mailer/incoming/incoming_handler.go]: but in the api, user with write permission can comment on locked issue. ref:
https://github.com/go-gitea/gitea/blob/9f919cf08339e8ec8f6b0b2a0dcafb3ea905fbec/routers/api/v1/repo/issue_comment.go#L360-L363

if use your logic, they can't. I wonder ...
then use ``||`` and ``&&`` but not use "(" is really hard to be understand for coder ...

Comment (wolfogre) [reply services/mailer/incoming/incoming_handler.go]: I see, could you please change it to `if issue.IsLocked && ...` too? It does make it easier to understand.

**Comment by a1012112796** [comment]
@KN4CK3R wonder why get a error when shutdown (tested by outlook)
image

**Comment by KN4CK3R** [comment]
> for others, read permission is enough

Is that correct? That does not sound good.

---

> @KN4CK3R wonder why get a error when shutdown (tested by outlook) image

Hm, I have never seen this message.

**Comment by a1012112796** [comment]
> > for others, read permission is enough
> 
> Is that correct? That does not sound good.

that's same with web ui and api, looks github has same logic also.

File: services/mailer/incoming/incoming_handler.go (hunk 1 above)
Comment (a1012112796) [reply services/mailer/incoming/incoming_handler.go]:
```suggestion
	if issue.IsLocked && !perm.CanWriteIssuesOrPulls(issue.IsPull) && !doer.IsAdmin {
```

this way?

Comment (wolfogre) [reply services/mailer/incoming/incoming_handler.go]: Yes, so it will be clear that this is to deal with locked issues.

Comment (delvh) [reply services/mailer/incoming/incoming_handler.go]:
```suggestion
	// Locked issues require write permissions
	if issue.IsLocked && !perm.CanWriteIssuesOrPulls(issue.IsPull) && !doer.IsAdmin {
```

Comment (delvh) [reply services/mailer/incoming/incoming_handler.go]: Wait, is `Read` really enough to `Write` an issue comment?
I can see a lot of problems waiting for us, i.e. when we add private issues…

Comment (lunny) [reply services/mailer/incoming/incoming_handler.go]: > 

Yes. If you have read permission of issue, you can create issue and post comments.

Comment (delvh) [reply services/mailer/incoming/incoming_handler.go]: As I said: That will certainly make implementing private issues (safely) a pain.

**Comment by KN4CK3R** [comment]
If the "write check" does not do what I think it was doing, we need another check to see if the user has access to the **repo**. Otherwise an user could reply to an older mail belonging to a now private repo.

**Comment by delvh** [comment]
↑ That's exactly what I mean, it produces a lot of problems.

**Comment by a1012112796** [comment]
read permission check has do it, isn't it?
> we need another check to see if the user has access to the **repo**.

File: services/mailer/incoming/incoming_handler.go (hunk 1 above)
Comment (a1012112796) [reply services/mailer/incoming/incoming_handler.go]: > As I said: That will certainly make implementing private issues (safely) a pain.

I think has been diffcult now ... , when implementing private issues. it's better to use a unique servcie fuction to create commnt for both web ui, api and incoming mail...

**Comment by a1012112796** [comment]
> read permission check has do it, isn't it?
> 
> > we need another check to see if the user has access to the **repo**.

**Review by wolfogre — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
make LG-TM work

**Pull request merged**